proposals/28-deal-flow-e2e-tests.md

```diff
@@ -0,0 +1,192 @@
+
+Explain why this rating
+-->
+
+```
+ICE Score = (Impact + Confidence + Ease) / 3
+ICE Score = (7 + 7 + 8) / 3
+ICE Score = 7.333
+```
+
+**Impact 7** because the project touches on the following areas:
+
+* User Aquuisition - Doesn't Abandon.
+* Activation - Makes a deal first time.
+* Retention - Reducing bad experiences that force the user to try something else.
+* Revenue - Generates revenue.
+
+**Confidence 7** because these will generate test results and benchmarking data.
+
+**Ease 8** an estimate that this will take 3 weeks to get up and running. The time is reltively short because there's no changes to Lotus that need to be made, it can be created as a separate greenfield project.
+
+## Project definition
+#### Brief plan of attack
+<!--Briefly describe the milestones/steps/work needed for this project-->
+
+1. Identify test scenarios.
+    * See the "Lotus Release Checklist" for some desired E2E tests to include.
+1. Identify tests that will generate benchmarking data.
+1. Decide on runner (BYO or testground?).
+1. Build a runner/learn a testground.
+1. Write the tests.
+    * Observe the pain points.
+    * Experience the bugs and open issues.
+    * Note the bottlenecks and open issues.
+    * Ensure tests with obvious bottlenecks generate benchmarking data.
+1. Integrate the suite (CI in Lotus/run nightly etc.).
+1. Collect/graph the metrics.
+1. Create new project(s) to streamline the flow.
+
+#### What does done look like?
+<!--
+_What specific deliverables should completed to consider this project done?_
+-->
+
+1. E2E test suite exists.
+1. E2E tests are run frequently.
+1. Metrics are graphed for other projects to use.
+
+####  What does success look like?
+<!--
+_Success means impact. How will we know we did the right thing?_
+
+Provide success criteria. These might include particular metrics, desired changes in the types of bug reports being filed, desired changes in qualitative user feedback (measured via surveys, etc), etc.
+-->
+
+1. One or more bugs are found in the deal flow and fixed in Lotus.
+1. The E2E tests catch a bug that would not have been otherwise caught until found in production.
+1. Another w3dt project uses metrics from these E2E tests to define it's success.
+1. Andrew Nesbitt is able to successfully negotiate a storage deal.
+1. Over time, performance for the collected metrics improves.
+
+#### Counterpoints &amp; pre-mortem
+<!--
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+-->
+
+If the deal flow changes before the project is complete then there will be additional work to do to update the tests either at the end or mid development.
+
+If the tests are "flakey" then their results will not be trusted and they may become abandoned.
```

Comment:
Agreed.  Should an exit criteria be that the tests can pass X times (e.g., 100) with a pass rate of Y (e.g., 99)?

Reply:
I'm somewhat against this because it can mask real issues. In my experience it is almost always because of a flakey test or bug in the code that a test fails, not for other reasons. We want to fix both of those two things, not ignore them.

When there's a legitimate need for this I've seen test frameworks provide a e.g. `retries(3)` function where the test can fail up to 2 times.